internal/pkg/term/syncbuffer/syncbuffer.go

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Package syncbuffer provides a synchronized buffer to store and access logs from multiple goroutines.
+package syncbuffer
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strings"
+	"sync"
+
+	"golang.org/x/term"
+)
+
+// maxLogLines is the maximum number of lines to display in the terminal.
+const maxLogLines = 5
+
+// fileWriter is the interface to write to a file.
+type fileWriter interface {
+	io.Writer
+}
+
+// SyncBuffer is a synchronized buffer that can be used to store output data and coordinate between multiple goroutines.
+type SyncBuffer struct {
+	BufMu sync.Mutex   // BufMu is a mutex that can be used to protect access to the buffer.
+	Buf   bytes.Buffer // Buf is the buffer that stores the data.
+
+	Done chan struct{} // Done is a channel that can be used to signal when the operations are complete.
+}
+
+// Write appends the given bytes to the buffer.
+func (b *SyncBuffer) Write(p []byte) (n int, err error) {
+	b.BufMu.Lock()
+	defer b.BufMu.Unlock()
+	return b.Buf.Write(p)
+}
+
+// Strings returns an empty slice if the buffer is empty.
+// Otherwise, it returns a slice of all the lines stored in the buffer.
+func (b *SyncBuffer) strings() []string {
+	b.BufMu.Lock()
+	defer b.BufMu.Unlock()
+
+	// Split the buffer bytes into lines.
+	lines := strings.Split(strings.TrimSpace(b.Buf.String()), "\n")
+
+	return lines
+}
+
+// IsDone returns true if the Done channel has been closed, otherwise return false.
+func (b *SyncBuffer) IsDone() bool {
+	select {
+	case <-b.Done:
+		return true
+	default:
+		return false
+	}
+}
+
+// MarkDone closes the Done channel.
+func (b *SyncBuffer) MarkDone() {
+	close(b.Done)
+}
+
+// TermPrinter is a printer to display logs in the terminal.
+type TermPrinter struct {
+	Term             fileWriter
+	Buf              *SyncBuffer
+	PrevWrittenLines int
+	TermWidth        int
+}
+
+// NewTermPrinter returns a new instance of TermPrinter that writes logs to the given file writer and reads logs from a new synchronized buffer.
+func NewTermPrinter(fw fileWriter) *TermPrinter {
+	return &TermPrinter{
+		Term: fw,
+		Buf: &SyncBuffer{
+			Done: make(chan struct{}),
+		},
+	}
+}
+
+// PrintLastFiveLines prints the label and the last five lines of logs to the termPrinter fileWriter.
+func (t *TermPrinter) PrintLastFiveLines() {
+	logs := t.Buf.strings()
+	outputLogs := t.lastFiveLogLines(logs)
+	if len(outputLogs) > 0 {
+		fmt.Fprintln(t.Term, logs[0])
+		for _, logLine := range outputLogs {
+			fmt.Fprintln(t.Term, logLine)
+		}
+		writtenLines := t.numLines(append(outputLogs[:], logs[0]))
+		t.PrevWrittenLines = writtenLines
+	}
+}
+
+// lastFiveLogLines returns the last five lines of the given logs, or all the logs if there are less than five lines.
+func (t *TermPrinter) lastFiveLogLines(logs []string) [maxLogLines]string {
+	// Determine the start and end index to extract last 5 lines
+	start := 1
```

Comment:
consider passing `logs[1:]` as input to this method so that we don't have to do this unusual starting position for the `lastNLines`